tests/run-pass/concurrency/sync.rs

```diff
@@ -0,0 +1,274 @@
+// ignore-windows: Concurrency on Windows is not supported yet.
+// compile-flags: -Zmiri-disable-isolation
+
+use std::sync::mpsc::{channel, sync_channel};
+use std::sync::{Arc, Barrier, Condvar, Mutex, Once, RwLock};
+use std::thread;
+use std::time::{Duration, Instant};
+
+// Check if Rust barriers are working.
+
+/// This test is taken from the Rust documentation.
+fn check_barriers() {
+    let mut handles = Vec::with_capacity(10);
+    let barrier = Arc::new(Barrier::new(10));
+    for _ in 0..10 {
+        let c = barrier.clone();
+        // The same messages will be printed together.
+        // You will NOT see any interleaving.
+        handles.push(thread::spawn(move || {
+            println!("before wait");
+            c.wait();
+            println!("after wait");
+        }));
+    }
+    // Wait for other threads to finish.
+    for handle in handles {
+        handle.join().unwrap();
+    }
+}
+
+// Check if Rust conditional variables are working.
+
+/// The test taken from the Rust documentation.
+fn check_conditional_variables_notify_one() {
+    let pair = Arc::new((Mutex::new(false), Condvar::new()));
+    let pair2 = pair.clone();
+
+    // Inside of our lock, spawn a new thread, and then wait for it to start.
+    thread::spawn(move || {
+        let (lock, cvar) = &*pair2;
```

Comment:
Maybe add a "yield" here to make sure we really get the main thread to block before coming here?